pkg/ignition/ignition.go

```diff
@@ -0,0 +1,118 @@
+	// CloudProviderOption is the kubelet CLI option for cloud provider
+	CloudProviderOption = "cloud-provider"
+)
+
+var (
+	// renderedWorker holds the contents of the cluster's worker node ignition file
+	renderedWorker Data
+)
+
+// Data holds the fully contents of an ignition file
+type Data struct {
+	ignCfgTypes.Config
+	// KubeletUnit is the kubelet systemd service spec in ignition file
+	KubeletUnit ignCfgTypes.Unit
+}
+
+// Initialize populates the RenderedWorker global variable with the worker ignition file data.
+func Initialize(c client.Client) error {
+	log := ctrl.Log.WithName("ignition")
+
+	mcList := &mcfg.MachineConfigList{}
+	err := c.List(context.TODO(), mcList)
+	if err != nil {
+		return err
+	}
+	renderedWorkerPrefix := "rendered-worker-"
+	var rawIgnition []byte
+	for _, mc := range mcList.Items {
+		if strings.HasPrefix(mc.Name, renderedWorkerPrefix) {
+			rawIgnition = mc.Spec.Config.Raw
+			log.Info("Received ignition spec", "MachineConfig", mc.Name)
+			break
+		}
+	}
+
+	configuration, report, err := ignCfg.Parse(rawIgnition)
+	if err != nil || report.IsFatal() {
+		return errors.Errorf("failed to parse Ign spec config: %v\nReport: %v", err, report)
+	}
+	renderedWorker.Config = configuration
+	log.Info("parsed", "version", renderedWorker.Config.Ignition.Version)
+
+	for _, unit := range configuration.Systemd.Units {
+		if unit.Name == kubeletSystemdName {
+			renderedWorker.KubeletUnit = unit
+			break
+		}
+	}
+	if renderedWorker.KubeletUnit.Contents == nil {
+		return errors.Errorf("ignition missing kubelet systemd unit file")
+	}
+	log.Info("Extracted", "kubelet systemd unit", renderedWorker.KubeletUnit)
+	return nil
+}
+
+// GetKubeletArgs returns a set of arguments for kubelet.exe, as specified in the ignition file
+func GetKubeletArgs() (map[string]string, error) {
+	if renderedWorker.KubeletUnit.Contents == nil {
+		return nil, errors.New("kubelet unit contents nil")
+	}
+	argsFromIgnition, err := parseKubeletArgs(*renderedWorker.KubeletUnit.Contents)
+	if err != nil {
+		return nil, errors.Wrap(err, "error parsing kubelet systemd unit args")
+	}
+	return argsFromIgnition, nil
+}
+
+//parseKubeletArgs parses a systemd unit file, returning the kubelet args WMCO is interested in
+func parseKubeletArgs(unitContents string) (map[string]string, error) {
+	log := ctrl.Log.WithName("ignition")
+	log.Info("parsing args", "unitContents", unitContents)
+	windowsArgs := []string{CloudProviderOption, CloudConfigOption}
+	execSplit := strings.SplitN(unitContents, "ExecStart=", 2)
+	if len(execSplit) != 2 {
+		return nil, fmt.Errorf("unit missing ExecStart")
+	}
+	cmdEndSplit := strings.SplitN(execSplit[1], "\n\n", 2)
+	argumentSplit := strings.Split(cmdEndSplit[0], "\\\n")
+	kubeletArgs := make(map[string]string)
+	for _, arg := range argumentSplit[1:] {
+		arg = strings.TrimSpace(arg)
+		arg = strings.TrimPrefix(arg, "--")
+		keyValue := strings.SplitN(arg, "=", 2)
+		if len(keyValue) != 2 {
```

Comment:
Shouldn't we also include these?

Reply:
Added comment explaining why thats not needed